api/src/main/java/org/eclipse/microprofile/config/Config.java

```diff
@@ -139,7 +139,7 @@
 
     /**
      * Return all of the currently registered {@linkplain ConfigSource configuration sources} for this configuration.
-     * The returned sources will be sorted by descending ordinal value and name.
+     * The returned sources will be sorted by descending ordinal value and name, which can be iterated in a thread-safe manner.
```

Comment:
This necessarily means that for any given implementation either (a) the returned sources must be a snapshot-in-time (i.e. not changing "underneath" you as you iterate) or (b) the returned sources may only grow, not shrink.  In either case the assumed behavior must be made specific in the specification language.  I think this (still) relates to #371: is the return value and behavior of this method intended to describe a set of objects whose cardinality is unchanging, or is it intended to describe a set of objects whose cardinality can only grow, not shrink?

Reply:
I think I'd lean towards unchanging/immutable.

Not sure what @Emily-Jiang, @radcortez, or @dmlloyd think?

It should be easy enough to get a minor wording change in before 1.4

Reply:
Unchanging or immutable can have two meanings here and while I know I sound pedantic I really, really, really, really think this is important:

1. Unchanging with respect to cardinality, but not with respect to content.  (At time 0, the returned set has three elements in it.  They are A, B, and C.  At time 1, the returned set continues to have three elements in it (its cardinality is unchanging).  However now it has A, D and Q.)
2. Unchanging with respect to cardinality _and_ content.  (At times 0 and 1 and _n_, the returned set has three elements in it which are A, B and C.)

(There's even more nuance here about whether A, B and C can have changing properties of their own or not; I'm tabling that for now, but that's _also_ important.)

Reply:
Completely fair and understandable.

I'd lean towards option 2, as I think we want to promote the notion that once a `Config` is created it's not possible to add additional `ConfigSource` instances to change the config content. For that, you need to recreate `Config`, but I could be reading too much into the spec there.

Reply:
Naïvely I'd assume option 2 was intended as well, but [there is a section on dynamic `ConfigSource`s](https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc#custom-configsources-via-configsourceprovider) that _also_ doesn't say anything one way or another about the mutability of cardinality or content of `ConfigSource`s provided in this manner.  I'm still feeling like there are a lot of assumptions and implications in a set of documents that should be much more prescriptive.  I understand this is an iterative process but I want to ensure that these assumptions and implications remain in the spotlight as the effort moves forward.

Reply:
Ah yeah, I think we tried to cleanup the dynamic config source issue but didn't succeed.

And I appreciate you continuing to raise them as they need to be resolved

Reply:
I believe the initial thought was to alway have an unchanging set of sources both for content and cardinality. At least, this is what I would expect, since you are not able to add additional ConfigSources after a Config is created.

I think that the ability to dynamically reload or change ConfigSources should not be left unspecified for the implementor to choose. This is to big of a decision and would only cause confusion how this works between users.

Reply:
Respectfully, you believe and think this, and Ken assumes it, and Emily and Mark indicated at one point that it shouldn't be specified, which means that it actually would be possible…and I'm pointing out with this comment thread that for reasons enumerated above this behavior _must_ be _explicitly_ specified, not believed, assumed, indicated or thought.  So I personally agree with you but my point is none of this is spelled out, and it must be.

Reply:
Ah, but I completely agree with you. It should be specified one way or another.